crates/uv/src/commands/tool/run.rs

```diff
@@ -362,52 +305,163 @@ async fn show_help(
     Ok(())
 }
 
-/// Display a warning if an executable is not provided by package.
-///
-/// If found in a dependency of the requested package instead of the requested package itself, we will hint to use that instead.
-fn warn_executable_not_provided_by_package(
-    executable: &str,
-    from_package: &PackageName,
-    site_packages: &SitePackages,
+struct ExecutableProviderPackages<'a> {
+    /// The requested executable for the command
+    executable: &'a str,
+    /// The package from which the executable is expected to come from
+    from: &'a ToolRequirement,
+    /// The packages in the Pythonenvironment the command will run it
+    site_packages: &'a SitePackages,
+    /// The matching packages
+    packages: Vec<InstalledDist>,
     invocation_source: ToolRunCommand,
-) {
-    let packages = matching_packages(executable, site_packages);
-    if !packages
-        .iter()
-        .any(|package| package.name() == from_package)
-    {
+}
+
+impl<'a> ExecutableProviderPackages<'a> {
+    fn new(
+        executable: &'a str,
+        from: &'a ToolRequirement,
+        site_packages: &'a SitePackages,
+        invocation_source: ToolRunCommand,
+    ) -> Self {
+        let packages = matching_packages(executable, site_packages);
+        ExecutableProviderPackages {
+            executable,
+            from,
+            site_packages,
+            packages,
+            invocation_source,
+        }
+    }
+
+    fn not_from_expected(&self) -> bool {
+        let from = match self.from {
+            // Nothing to do for Python
+            ToolRequirement::Python => return false,
+            ToolRequirement::Package { requirement, .. } => requirement,
+        };
```

Comment:
I'm not super pleased with this. It'd be nice to just avoid this by construction.

Reply:
Fixed in https://github.com/astral-sh/uv/pull/11603/commits/913dd90ad41683532f8c961b14864055483a04fe